cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,279 @@
+	targetFirewall := fm.fwCache.firewall
+	fm.fwCache.mu.RUnlock()
+
+	fwSvc := fm.client.Firewalls
+
+	fm.fwCache.mu.RLock()
+	isEqual := cmp.Equal(targetFirewall.InboundRules, svcInboundRules)
+	fm.fwCache.mu.RUnlock()
+
+	if !isEqual {
+		currentFirewall, err := fm.Get(ctx)
+		if err != nil {
+			// Log error then create firewall and update cache, since firewall does not exist.
+			klog.Errorf("failed to get firewall state: %s", err)
+			currentFirewall, err = fm.createFirewall(ctx, fwSvc, svcInboundRules)
+			if err != nil {
+				return err
+			}
+			fm.updateCache(currentFirewall)
+			return nil
+		} else {
+			// Update api firewall inbound rules with the expected inbound rules then update cache.
+			fr := &godo.FirewallRequest{
+				Name:         fm.workerFirewallName,
+				InboundRules: svcInboundRules,
+			}
+			currentFirewall, resp, err := fwSvc.Update(ctx, currentFirewall.ID, fr)
+			if err != nil {
+				if resp == nil || resp.StatusCode != http.StatusNotFound {
+					return fmt.Errorf("failed to update firewall: %v", err)
+				}
+				// firewall does not exist, so we should create a new one.
+				currentFirewall, err = fm.createFirewall(ctx, fwSvc, svcInboundRules)
+				if err != nil {
+					return fmt.Errorf("failed to add new firewall inbound rules: %s", err)
+				}
+			}
+			// if the firewall api update has succeeded, proceed to update the local firewall
+			// cache.
+			fm.updateCache(currentFirewall)
+		}
+	}
+	return nil
+}
+
+func (fm *firewallManagerOp) createFirewall(ctx context.Context, fwSvc godo.FirewallsService, svcInboundRules []godo.InboundRule) (*godo.Firewall, error) {
+	fr := &godo.FirewallRequest{
+		Name:         fm.workerFirewallName,
+		InboundRules: svcInboundRules,
+		// Tags:          currentFirewall.Tags,
+	}
+	currentFirewall, _, err := fwSvc.Create(ctx, fr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new firewall: %v", err)
+	}
+
+	return currentFirewall, nil
+}
+
+func (fm *firewallManagerOp) updateCache(currentFirewall *godo.Firewall) {
+	fm.fwCache.mu.RLock()
+	defer fm.fwCache.mu.RUnlock()
+	fm.fwCache.firewall = currentFirewall
+}
+
+func (fc *FirewallController) ensureReconciledFirewall(ctx context.Context) error {
+	serviceList, err := fc.serviceLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to list services: %s", err)
+	}
+	inboundRules := fc.createInboundRules(serviceList)
+	fc.fwManager.Set(ctx, inboundRules)
+	return nil
+}
+
+func (fc *FirewallController) createInboundRules(serviceList []*v1.Service) []godo.InboundRule {
+	var nodePortInboundRules []godo.InboundRule
+	for _, svc := range serviceList {
+		if svc.Spec.Type == v1.ServiceTypeNodePort {
+			// this is a nodeport service so we should check for existing inbound rules on all ports.
+			for _, servicePort := range svc.Spec.Ports {
+				// In the odd case that a failure is asynchronous causing the NodePort to be set to zero.
+				if servicePort.NodePort == 0 {
+					continue
```

Comment:
Can you log a `klog.Warning` here? This should really not happen, but if it does we want to be able to see it.